zettarepl/zettarepl.py

```diff
@@ -132,17 +137,89 @@ def _replication_tasks_for_periodic_snapshot_tasks(self, replication_tasks, peri
 
         return result
 
-    def _run_replication_tasks(self, replication_tasks):
-        for transport, replication_tasks in self._transport_for_replication_tasks(replication_tasks):
-            remote_shell = self._get_shell(transport)
+    def _spawn_replication_tasks(self, replication_tasks):
+        with self.tasks_lock:
+            for replication_task in replication_tasks:
+                if replication_task in self.pending_tasks:
+                    logger.debug("Replication task %r is already pending")
+                    continue
+
+                if self._can_spawn_replication_task(replication_task):
+                    self._spawn_replication_task(replication_task)
+                else:
+                    logger.info("Replication task %r can't execute in parallel with already running tasks, "
+                                "delaying it", replication_task)
+                    notify(self.observer, ReplicationTaskScheduled(replication_task.id))
+                    self.pending_tasks.append(replication_task)
+
+            if not self.pending_tasks and not self.running_tasks:
+                self._spawn_retention()
+
+    def _can_spawn_replication_task(self, replication_task: ReplicationTask):
+        if self.retention_running:
+            return False
+
+        return all(self._replication_tasks_can_run_in_parallel(replication_task, t) for t in self.running_tasks)
 
-            push_replication_tasks, replication_tasks = bisect(self._is_push_replication_task, replication_tasks)
-            run_replication_tasks(self.local_shell, transport, remote_shell, push_replication_tasks, self.observer)
+    def _replication_tasks_can_run_in_parallel(self, t1: ReplicationTask, t2: ReplicationTask):
+        if t1.direction != t2.direction:
+            return True
+
+        if not are_same_host(t1.transport, t2.transport):
+            return True
+
+        return not is_child(t1.target_dataset, t2.target_dataset) and not is_child(t2.target_dataset, t1.target_dataset)
 
-            pull_replication_tasks, replication_tasks = bisect(self._is_pull_replication_task, replication_tasks)
-            run_replication_tasks(self.local_shell, transport, remote_shell, pull_replication_tasks, self.observer)
+    def _spawn_replication_task(self, replication_task):
+        self.running_tasks.append(replication_task)
+        threading.Thread(name=f"replication_task__{replication_task.id}",
+                         target=functools.partial(self._run_replication_task, replication_task)).start()
```

Comment:
Do we really want threads running without daemon mode?

Reply:
Yes, it's for `run --once` mode